components/script/dom/htmlmeterelement.rs

```diff
@@ -56,6 +76,98 @@ impl HTMLMeterElement {
             can_gc,
         )
     }
+
+    fn create_shadow_tree(&self, can_gc: CanGc) {
+        let document = self.owner_document();
+        let root = self
+            .upcast::<Element>()
+            .attach_shadow(
+                IsUserAgentWidget::Yes,
+                ShadowRootMode::Closed,
+                false,
+                SlotAssignmentMode::Manual,
+            )
+            .expect("Attaching UA shadow root failed");
+
+        let meter_value = HTMLDivElement::new(local_name!("div"), None, &document, None, can_gc);
+        root.upcast::<Node>()
+            .AppendChild(meter_value.upcast::<Node>())
+            .unwrap();
+
+        let _ = self.shadow_tree.borrow_mut().insert(ShadowTree {
+            meter_value: meter_value.as_traced(),
+        });
+        self.upcast::<Node>()
+            .dirty(crate::dom::node::NodeDamage::OtherNodeDamage);
+    }
+
+    fn shadow_tree(&self, can_gc: CanGc) -> Ref<'_, ShadowTree> {
+        if !self.upcast::<Element>().is_shadow_host() {
+            self.create_shadow_tree(can_gc);
+        }
+
+        Ref::filter_map(self.shadow_tree.borrow(), Option::as_ref)
+            .ok()
+            .expect("UA shadow tree was not created")
+    }
+
+    fn update_state(&self, can_gc: CanGc) {
+        let value = *self.Value();
```

Comment:
Might be a good idea to have internal methods for the attribute access(which would take the other attributes as arguments), which would be used to implement `HTMLMeterElementMethods` and  also here, so that here you wouldn't have to make redundant calls to the public methods(it would also be clearer how this matches the spec, because my first reaction was that the calls here are not made in the right order).

for example:

```
fn value(&self, min: Finite<f64>, max: Finite<f64>) -> Finite<f64>
fn min(&self) -> Finite<f64>
```

where

```
fn Value(&self) -> Finite<f64> {
    let min = self.min();
    let max = self.max();
    self.value(min, max)
}

Here again documenting the operation with the matching prose would be good.

Reply:
Hmm, all of this is probably inlined anyways and I don't see how it makes the code any clearer to be honest.

Reply:
The spec says specifically that the ordering is important because of how members are computed, and the code seems wrong when you first look at it; it only becomes clear that is is correct, but includes redundant calls, when you look at the implementation of the trait methods. So that means that if you separated those methods, and would call only the private ones in the right order  in `update_state`, the code would more obviously follow the spec. 

It is a small deal for this one specific part of the code, but compounded over a large codebase things like this make a big difference.